doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1613 @@
+  ...         if isinstance(m, Integer) and isinstance(n, Integer):
+  ...             return int(n % m == 0)
+  ...
+  ...         # For symbolic arguments, require m and n to be integer.
+  ...         # This is the better way to write this logic.
+  ...         if m.is_integer is False or n.is_integer is False:
+  ...             raise TypeError("m and n should be integers")
+  ```
+
+  This still disallows non-integer inputs as desired:
+
+  ```py
+  >>> divides(1.5, 1)
+  Traceback (most recent call last):
+  ...
+  TypeError: m and n should be integers
+  ```
+
+  But it does not fail in cases where the assumption is `None`:
+
+  ```py
+  >>> divides(2, (m**2 + m)/2)
+  divides(2, m**2/2 + m/2)
+  >>> _.subs(m, 2)
+  0
+  >>> n, m = symbols('n m') # Redefine n and m without the integer assumption
+  >>> divides(m, n)
+  divides(m, n)
+  ```
+
+  ```{note}
+
+  This rule of allowing `None` assumptions only applies to instances
+  where an exception would be raised, such as type checking an input domain.
+  In cases where simplifications or other operations are done, one should
+  treat a `None` assumption as meaning "can be either `True` or `False`" and
+  not perform an operation that might not be mathematically valid.
+
+  ```
+
+(custom-functions-assumptions)=
+### Assumptions
+
+The next thing we might want to define are the assumptions on our function.
+The assumptions system allows defining what mathematical properties your
+function has given its inputs, for example, "$f(x)$ is *positive* when $x$ is
+*real*."
+
+
+The [guide on the assumptions system](assumptions-guide) goes into the
+assumptions system in great detail. It is recommended to read through that
+guide first to understand what the different assumptions mean and how the
+assumptions system works.
+
+The simplest case is a function that always has a given assumption regardless
+of its input. In this case, you can define <code class="docutils
+literal notranslate"><span class="pre">is_*assumption*</span></code> directly
+on the class.
+
+For example, the `divides` function [defined above](custom-functions-divides)
+is always an integer, because its value is always either 0 or 1.
+
+```{note}
+
+From here on out in this guide, in the interest of space, we will omit the
+previous method definitions in the examples unless they are needed for the
+given example to work. There are [complete
+examples](custom-functions-complete-examples) with all methods included at the
+end of this guide.
+
+```
+
+```py
+>>> class divides(Function):
+...     is_integer = True
+...     is_negative = False
+```
+
+```py
+>>> divides(m, n).is_integer
+True
+>>> divides(m, n).is_nonnegative
+True
+```
+
+In general, however, the assumptions of a function depend on the assumptions
+of its inputs. In this case, you should define an <code class="docutils literal notranslate"><span
+class="pre">\_eval\_*assumption*</span></code> method.
```

Comment:
I was going to add an admonition here about not defining `is_assumption` as a @property method and using `_eval_is_assumption` instead, but I'm not actually sure if anything bad would happen if you do that, other than just being bad style. @oscarbenjamin do you know if anything particularly bad can happen if you do this? I tested it and it seems to actually work.

Reply:
It won't work with the assumptions cache and won't compute implications.
```
In [110]: exp(2)._assumptions
Out[110]: {}

In [111]: exp(2).is_zero
Out[111]: False

In [112]: exp(2)._assumptions
Out[112]: {'zero': False}

In [113]: exp(2).is_negative
Out[113]: False

In [114]: exp(2)._assumptions
Out[114]: 
{'zero': False,
 'finite': True,
 'infinite': False,
 'extended_negative': False,
 'negative': False}

In [115]: class A(Expr):
     ...:     @property
     ...:     def is_positive(self):
     ...:         return True
     ...: 

In [116]: a = A()

In [117]: a.is_positive
Out[117]: True

In [118]: a._assumptions
Out[118]: {}
```

Reply:
Ah you are right. For some reason I didn't notice this when I was playing with it before:

```
>>> class Test(Function):
...     @property
...     def is_zero(self):
...         return self.args[0].is_zero
>>> t = Test(0)
>>> t.is_zero
True
>>> t._assumptions
{}
>>> t.is_negative # This is not computed
>>> class Test2(Function):
...     def _eval_is_zero(self):
...         return self.args[0].is_zero
>>> t2 = Test2(0)
>>> t2.is_zero
True
>>> t2._assumptions
{'zero': True, 'negative': False, 'nonzero': False, 'extended_negative': False, 'integer': True, 'infinite': False, 'complex': True, 'odd': False, 'extended_positive': False, 'extended_nonpositive': True, 'algebraic': True, 'irrational': False, 'extended_real': True, 'nonnegative': True, 'imaginary': False, 'prime': False, 'composite': False, 'finite': True, 'transcendental': False, 'positive': False, 'extended_nonzero': False, 'even': True, 'real': True, 'noninteger': False, 'nonpositive': True, 'hermitian': True, 'extended_nonnegative': True, 'rational': True, 'commutative': True}
>>> t2.is_negative
False
```

Reply:
I wonder if we ought to try to detect and warn about this in the metaclass (as long as it wouldn't hurt performance).

Reply:
I have a proof of concept of that at https://github.com/sympy/sympy/pull/23640. However, there are a few issues with it that we'd need to figure out first. So for now, I've just included a warning in this doc about this. If that PR gets merged we can remove the warning, or at least tone it down.